src/net_processing.cpp

```diff
@@ -663,6 +671,35 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
         // This step done after final handshake
         CheckAndRequestExpeditedBlocks(pfrom);
     }
+    else if (strCommand == NetMsgType::XUPDATE)
+    {
+        CXVersionMessage xUpdate;
+        vRecv >> xUpdate;
+        // check for peer trying to change non-changeable key
+        for (auto entry : xUpdate.xmap)
+        {
+            auto iter = XVer::keytype.find(entry.first);
+            if (iter == XVer::keytype.end())
+            {
+                pfrom->fDisconnect = true;
+                LOG(NET, "ERROR: disconnecting - peer=%s attempting to update non-existing xversion value\n",
+                    pfrom->GetLogName());
+                return false;
+            }
+            else if (iter->second != "c")
+            {
+                pfrom->fDisconnect = true;
+                LOG(NET, "ERROR: disconnecting - peer=%s attempting to update non-changeable xversion value\n",
+                    pfrom->GetLogName());
+                return false;
+            }
+        }
+        // no issues? continue and actually update
+        for (auto entry : xUpdate.xmap)
+        {
+            pfrom->xState[entry.first] = xUpdate.xmap[entry.first];
```

Comment:
Looks good, though I now wonder whether you want to keep a separate `xUpdate` map or rather have a single xversion table and update the values that are changeable therein.

I suspect a single table will be easier to handle in the long run, as it is less cognitive load on the user of the table (which one do I need, again?), less code to deal with the table in general (as there is just one, I am talking about things like per-node get-accessors to this table) and also if you update a key from initial to changeable for a feature, it becomes easier to update the code. You won't have to update the table you refer to then, just change key type `initial` -> `changeable` and update the code to handle it (which you'd have to do in either scenario).

That said, and given such a potential upgrade path, maybe changes to keys that are initial should actually simply be *silently ignored* instead of leading to disconnects?

I know that this is different from what I proposed in chat, but with you bringing up an implementation prototype here, I think it is good to mull those things over carefully and I can't say I forethought everything here :-)

What's your thoughts on this?

Reply:
i dont see a problem with silently ignoring. i also dont see an issue with changing xState to simply keep track of what keys we allow changes to and still updating their values in xversion to address the "which table do i need again" issue you brought up. but i would like a secondary map to simply keep track of what keys we allow changes to vs searching through all keys every time,

Reply:
after your re-explanation in chat i will remove xstate and use only a single table